Source/Dafny/Substituter.cs

```diff
@@ -59,7 +59,7 @@ public class Substituter {
       } else if (expr is DisplayExpression) {
         DisplayExpression e = (DisplayExpression)expr;
         List<Expression> newElements = SubstituteExprList(e.Elements);
-        if (newElements != e.Elements) {
+        if (newElements != e.Elements || !Resolver.SubstType(expr.Type, typeMap).Equals(e.Type)) {
```

Comment:
```suggestion
        if (newElements != e.Elements || !Resolver.SubstType(e.Type, typeMap).Equals(e.Type)) {
```

It took me longer than necessary to figure out that `expr` and `e`` are the same object. Is there any reason not to use the same variable?

Also, Equals() has two variants, one that considers equality modulo subset types (the one you use), and the other that considers equality including subset types (with a second argument `, true`). I'm not sure, but I would have thought the second type of equality might have been appropriate. Could you let me know your reasoning for choosing one over the other?

Reply:
I should have used `e` there, thanks. for the latter I don't know about the other type of `Equals` are you suggesting to change it with `Equals(e.Type,true))` ?

Reply:
I'm not sure about that, but yes, that's my suggestion.